src/main/query_profiler.cpp

```diff
@@ -464,6 +473,64 @@ static string RenderTiming(double timing) {
 	return timing_s + "s";
 }
 
+string QueryProfiler::ToHTML() const {
+	std::stringstream str;
+	HTMLToStream(str);
+	return str.str();
+}
+
+void QueryProfiler::HTMLToStream(std::ostream &ss) const {
+	if (!IsEnabled()) {
+		ss << "Query profiling is disabled. Use 'PRAGMA enable_profiling;' to enable profiling!";
+		return;
+	}
+	// checking the tree to ensure the query is really empty
+	// the query string is empty when a logical plan is deserialized
+	if (query_info.query_name.empty() && !root) {
+		return;
+	}
+
+	// render the main operator tree
+	if (root) {
+		// Set the query elapsed time in the root node.
+		auto &info = root->GetProfilingInfo();
+		if (info.Enabled(MetricsType::OPERATOR_TIMING)) {
+			info.metrics[MetricsType::OPERATOR_TIMING] = main_query.Elapsed();
+		}
+		HTMLTreeRenderer renderer;
+		renderer.Render(*root, ss);
+	}
+}
+
+string QueryProfiler::ToGraphviz() const {
+	std::stringstream str;
+	GraphvizToStream(str);
+	return str.str();
+}
+
+void QueryProfiler::GraphvizToStream(std::ostream &ss) const {
```

Comment:
Same comment as above, can we avoid this duplication?

Reply:
Note: I also changed the call from `connection.cpp` to call into `query_profiler::ToString(ProfilerPrintFormat)`